src/lib/linux-api/src/capability.rs

```diff
@@ -0,0 +1,62 @@
+use linux_syscall::syscall;
+use linux_syscall::Result as LinuxSyscallResult;
+
+use crate::bindings;
+use crate::errno::Errno;
+
+pub const LINUX_CAPABILITY_VERSION_3: u32 = bindings::LINUX__LINUX_CAPABILITY_VERSION_3;
+
+#[allow(non_camel_case_types)]
+pub type user_cap_header = __user_cap_header_struct;
+#[allow(non_camel_case_types)]
+pub type user_cap_data = __user_cap_data_struct;
+
+// Somehow this is not automatically generated in bindings.rs
+#[repr(C)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct __user_cap_header_struct {
+    pub version: bindings::linux___u32,
+    pub pid: ::core::ffi::c_int,
+}
+
+// Somehow this is not automatically generated in bindings.rs
+#[repr(C)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct __user_cap_data_struct {
+    pub effective: bindings::linux___u32,
+    pub permitted: bindings::linux___u32,
+    pub inheritable: bindings::linux___u32,
+}
+
+unsafe fn capget_raw(hdrp: *const user_cap_header, datap: *mut user_cap_data) -> Result<(), Errno> {
```

Comment:
I don't quite understand this comment. What is `SYSCALL_DEFINE2`? `user_cap_header` and `user_cap_data` are just `__user_cap_header_struct` and `__user_cap_data_struct`. I changed the name because the Linux names look quite bad; they have the `__` prefix and the `_struct` suffix.

Does this one https://github.com/shadow/shadow/pull/3198/commits/1de8496c7e63563f36e2f9dd5b72e10d2ff0d137 look good for you?

Reply:
Ah sorry I should have been more clear. The `SYSCALL_DEFINE2` is the actual linux syscall implementation.

https://github.com/torvalds/linux/blob/b401b621758e46812da61fa58a67c3fd8d91de0d/kernel/capability.c#L141

Linux casts the capget syscall arguments as `header: cap_user_header_t` and `dataptr: cap_user_data_t`. Since your wrapper in the linux-api library uses `header: *const user_cap_header` and `dataptr: *mut user_cap_data`, I was looking for a comment that explains the relationship between those types. Typically we'd want to use the exact same type in our wrapper as what Linux uses, but it's fine in this case since they are the same because `linux/capability.h` defines:

```c
typedef struct __user_cap_header_struct {
	__u32 version;
	int pid;
} __user *cap_user_header_t;

struct __user_cap_data_struct {
        __u32 effective;
        __u32 permitted;
        __u32 inheritable;
};
typedef struct __user_cap_data_struct __user *cap_user_data_t;
```

So your `*user_cap_header` is a typedef of `*__user_cap_header_struct`, which is also typedefed as `cap_user_header_t`.

> Does this one 1de8496 look good for you?

Looks good to me, but the `SYSCALL_DEFINE2` lines should be an exact copy-paste of the Linux source code. So just needs the following changes:

https://github.com/torvalds/linux/blob/b401b621758e46812da61fa58a67c3fd8d91de0d/kernel/capability.c#L141

```diff
-SYSCALL_DEFINE2(capget, cap_user_header_t, hdrp, cap_user_data_t, datap)
+SYSCALL_DEFINE2(capget, cap_user_header_t, header, cap_user_data_t, dataptr)
```

and:

https://github.com/torvalds/linux/blob/b401b621758e46812da61fa58a67c3fd8d91de0d/kernel/capability.c#L220

```diff
-SYSCALL_DEFINE2(capset, cap_user_header_t, hdrp, cap_user_data_t, datap)
+SYSCALL_DEFINE2(capset, cap_user_header_t, header, const cap_user_data_t, data)
```